File: src/main/java/com/fengsheng/skill/MiaoShou.java
package com.fengsheng.skill;

import com.fengsheng.*;
import com.fengsheng.card.Card;
import com.fengsheng.phase.FightPhaseIdle;
import com.fengsheng.protos.Role;
import com.google.protobuf.GeneratedMessageV3;
import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

/**
 * 阿芙罗拉技能【妙手】：争夺阶段，你可以翻开此角色牌，然后弃置待接收情报，并查看一名角色的手牌和情报区，从中选择一张牌作为待收情报，面朝上移至一名角色的面前。
 */
public class MiaoShou extends AbstractSkill implements ActiveSkill {
    private static final Logger log = Logger.getLogger(MiaoShou.class);

    @Override
    public SkillId getSkillId() {
        return SkillId.MIAO_SHOU;
    }

    @Override
    public void executeProtocol(Game g, Player r, GeneratedMessageV3 message) {
        if (!(g.getFsm() instanceof FightPhaseIdle fsm) || r != fsm.whoseFightTurn) {
            log.error("现在不是发动[妙手]的时机");
            return;
        }
        if (r.isRoleFaceUp()) {
            log.error("你现在正面朝上，不能发动[妙手]");
            return;
        }
        var pb = (Role.skill_miao_shou_a_tos) message;
        if ((r instanceof HumanPlayer humanPlayer) && !humanPlayer.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        if (pb.getTargetPlayerId() < 0 || pb.getTargetPlayerId() >= g.getPlayers().length) {
            log.error("目标错误");
            return;
        }
        Player target = g.getPlayers()[r.getAbstractLocation(pb.getTargetPlayerId())];
        if (!target.isAlive()) {
            log.error("目标已死亡");
            return;
        }
        if (target.getCards().isEmpty() && target.getMessageCards().isEmpty()) {
            log.error("目标没有手牌，也没有情报牌");
            return;
        }
        r.incrSeq();
        r.addSkillUseCount(getSkillId());
        g.playerSetRoleFaceUp(r, true);
        g.getDeck().discard(fsm.messageCard);
        g.resolve(new executeMiaoShou(fsm, r, target));
    }

    private record executeMiaoShou(FightPhaseIdle fsm, Player r, Player target) implements WaitingFsm {
        @Override
        public ResolveResult resolve() {
            final Game g = r.getGame();
            log.info(r + "对" + target + "发动了[妙手]");
            for (Player p : g.getPlayers()) {
                if (p instanceof HumanPlayer player) {
                    var builder = Role.skill_miao_shou_a_toc.newBuilder();
                    builder.setPlayerId(player.getAlternativeLocation(r.location()));
                    builder.setTargetPlayerId(player.getAlternativeLocation(target.location()));
                    builder.setWaitingSecond(20);
                    if (player == r) {
                        for (Card card : target.getCards().values())
                            builder.addCards(card.toPbCard());
                        final int seq2 = player.getSeq();
                        builder.setSeq(seq2);
                        player.setTimeout(GameExecutor.post(g, () -> {
                            if (player.checkSeq(seq2)) {
                                for (Card card : target.getCards().values()) {
                                    g.tryContinueResolveProtocol(r, Role.skill_miao_shou_b_tos.newBuilder()
                                            .setCardId(card.getId()).setTargetPlayerId(0).setSeq(seq2).build());
                                    return;
                                }
                                for (Card card : target.getMessageCards().values()) {
                                    g.tryContinueResolveProtocol(r, Role.skill_miao_shou_b_tos.newBuilder()
                                            .setMessageCardId(card.getId()).setTargetPlayerId(0).setSeq(seq2).build());
                                    return;
                                }
                            }
                        }, player.getWaitSeconds(builder.getWaitingSecond() + 2), TimeUnit.SECONDS));
                    }
                    player.send(builder.build());
                }
            }
            if (r instanceof RobotPlayer) {
                GameExecutor.post(g, () -> {
                    for (Card card : target.getCards().values()) {
                        g.tryContinueResolveProtocol(r, Role.skill_miao_shou_b_tos.newBuilder()
                                .setCardId(card.getId()).setTargetPlayerId(0).build());
                        return;
                    }
                    for (Card card : target.getMessageCards().values()) {
                        g.tryContinueResolveProtocol(r, Role.skill_miao_shou_b_tos.newBuilder()
                                .setMessageCardId(card.getId()).setTargetPlayerId(0).build());
                        return;
                    }
                }, 2, TimeUnit.SECONDS);
            }
            return null;
        }

        @Override
        public ResolveResult resolveProtocol(Player player, GeneratedMessageV3 message) {
            if (player != r) {
                log.error("不是你发技能的时机");
                return null;
            }
            if (!(message instanceof Role.skill_miao_shou_b_tos pb)) {
                log.error("错误的协议");
                return null;
            }
            Game g = r.getGame();
            if (r instanceof HumanPlayer humanPlayer && !humanPlayer.checkSeq(pb.getSeq())) {
                log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
                return null;
            }
            if (pb.getCardId() != 0 && pb.getMessageCardId() != 0) {
                log.error("只能选择手牌或情报其中之一");
                return null;
            }
            if (pb.getTargetPlayerId() < 0 || pb.getTargetPlayerId() >= g.getPlayers().length) {
                log.error("目标错误");
                return null;
            }
            Player target2 = g.getPlayers()[r.getAbstractLocation(pb.getTargetPlayerId())];
            if (!target2.isAlive()) {
                log.error("目标已死亡");
                return null;
            }
            Card card;
            if (pb.getCardId() == 0 && pb.getMessageCardId() == 0) {
                log.error("必须选择一张手牌或情报");
                return null;
            } else if (pb.getMessageCardId() == 0) {
                card = target.deleteCard(pb.getCardId());
                if (card == null) {
                    log.error("没有这张牌");
                    return null;
                }
            } else {
                card = target.deleteMessageCard(pb.getMessageCardId());
                if (card == null) {
                    log.error("没有这张牌");
                    return null;
                }
            }
            r.incrSeq();
            log.info(r + "将" + card + "作为情报，面朝上移至" + target2 + "的面前");
            fsm.messageCard = card;
            fsm.inFrontOfWhom = target2;
            fsm.whoseFightTurn = fsm.inFrontOfWhom;
            fsm.isMessageCardFaceUp = true;
            for (Player p : g.getPlayers()) {
                if (p instanceof HumanPlayer player1) {
                    var builder = Role.skill_miao_shou_b_toc.newBuilder();
                    builder.setPlayerId(player1.getAlternativeLocation(r.location()));
                    builder.setFromPlayerId(player1.getAlternativeLocation(target.location()));
                    builder.setTargetPlayerId(player1.getAlternativeLocation(target2.location()));
                    if (pb.getCardId() != 0) builder.setCard(card.toPbCard());
                    else builder.setMessageCardId(card.getId());
                    player1.send(builder.build());
                }
            }
            return new ResolveResult(fsm, true);
        }
    }

    public static boolean ai(FightPhaseIdle e, final ActiveSkill skill) {
        Player player = e.whoseFightTurn;
        if (player.isRoleFaceUp())
            return false;
        List<Player> players = new ArrayList<>();
        for (Player p : player.getGame().getPlayers()) {
            if (p.isAlive() && (!p.getCards().isEmpty() || !p.getMessageCards().isEmpty())) {
                players.add(p);
            }
        }
        if (players.isEmpty()) return false;
        if (ThreadLocalRandom.current().nextInt(players.size()) != 0)
            return false;
        Player p = players.get(ThreadLocalRandom.current().nextInt(players.size()));
        GameExecutor.post(player.getGame(), () -> skill.executeProtocol(
                player.getGame(), player, Role.skill_miao_shou_a_tos.newBuilder().setTargetPlayerId(p.location()).build()
        ), 2, TimeUnit.SECONDS);
        return true;
    }
}

File: src/main/kotlin/skill/MiaoShou.kt
package com.fengsheng.skill

import com.fengsheng.*
import com.fengsheng.RobotPlayer.Companion.sortCards
import com.fengsheng.card.Card
import com.fengsheng.phase.FightPhaseIdle
import com.fengsheng.phase.NextTurn
import com.fengsheng.protos.Common.color.Black
import com.fengsheng.protos.Role.skill_miao_shou_a_tos
import com.fengsheng.protos.Role.skill_miao_shou_b_tos
import com.fengsheng.protos.skillMiaoShouAToc
import com.fengsheng.protos.skillMiaoShouATos
import com.fengsheng.protos.skillMiaoShouBToc
import com.fengsheng.protos.skillMiaoShouBTos
import com.google.protobuf.GeneratedMessage
import org.apache.logging.log4j.kotlin.logger
import java.util.concurrent.TimeUnit

/**
 * 阿芙罗拉技能【妙手】：争夺阶段，你可以翻开此角色牌，然后弃置待接收情报，并查看一名角色的手牌和情报区，从中选择一张牌作为待收情报，面朝上移至一名角色的面前。
 */
class MiaoShou : ActiveSkill {
    override val skillId = SkillId.MIAO_SHOU

    override val isInitialSkill = true

    override fun canUse(fightPhase: FightPhaseIdle, r: Player): Boolean = !r.roleFaceUp

    override fun executeProtocol(g: Game, r: Player, message: GeneratedMessage) {
        val fsm = g.fsm as? FightPhaseIdle
        if (r !== fsm?.whoseFightTurn) {
            logger.error("现在不是发动[妙手]的时机")
            r.sendErrorMessage("现在不是发动[妙手]的时机")
            return
        }
        if (r.roleFaceUp) {
            logger.error("你现在正面朝上，不能发动[妙手]")
            r.sendErrorMessage("你现在正面朝上，不能发动[妙手]")
            return
        }
        val pb = message as skill_miao_shou_a_tos
        if (r is HumanPlayer && !r.checkSeq(pb.seq)) {
            logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${pb.seq}")
            r.sendErrorMessage("操作太晚了")
            return
        }
        if (pb.targetPlayerId < 0 || pb.targetPlayerId >= g.players.size) {
            logger.error("目标错误")
            r.sendErrorMessage("目标错误")
            return
        }
        val target = g.players[r.getAbstractLocation(pb.targetPlayerId)]!!
        if (!target.alive) {
            logger.error("目标已死亡")
            r.sendErrorMessage("目标已死亡")
            return
        }
        r.incrSeq()
        r.addSkillUseCount(skillId)
        g.playerSetRoleFaceUp(r, true)
        g.deck.discard(fsm.messageCard)
        if (target.cards.isEmpty() && target.messageCards.isEmpty()) {
            logger.info("${target}没有手牌，也没有情报，回合结束")
            g.resolve(NextTurn(fsm.whoseTurn))
            return
        }
        r.weiBiFailRate = 0
        g.resolve(ExecuteMiaoShou(fsm, r, target))
    }

    private data class ExecuteMiaoShou(val fsm: FightPhaseIdle, val r: Player, val target: Player) : WaitingFsm {
        override val whoseTurn: Player
            get() = fsm.whoseTurn

        override fun resolve(): ResolveResult? {
            val g = r.game!!
            logger.info("${r}对${target}发动了[妙手]")
            g.players.send { p ->
                skillMiaoShouAToc {
                    playerId = p.getAlternativeLocation(r.location)
                    targetPlayerId = p.getAlternativeLocation(target.location)
                    waitingSecond = Config.WaitSecond
                    messageCard = fsm.messageCard.toPbCard()
                    if (p === r) {
                        target.cards.forEach { cards.add(it.toPbCard()) }
                        val seq2 = p.seq
                        seq = seq2
                        p.timeout = GameExecutor.post(g, {
                            if (p.checkSeq(seq2)) {
                                g.tryContinueResolveProtocol(r, skillMiaoShouBTos {
                                    cardId = target.cards.firstOrNull()?.id ?: 0
                                    if (cardId == 0)
                                        messageCardId = target.messageCards.firstOrNull()?.id ?: 0
                                    targetPlayerId = 0
                                    seq = seq2
                                })
                            }
                        }, p.getWaitSeconds(waitingSecond + 2).toLong(), TimeUnit.SECONDS)
                    }
                }
            }
            if (r is RobotPlayer) {
                GameExecutor.post(g, {
                    g.tryContinueResolveProtocol(r, skillMiaoShouBTos {
                        var value = Int.MIN_VALUE
                        val players = g.players.filterNotNull().filter { it.alive }.shuffled()
                        val handCardScore = if (target.isEnemy(r)) 10 else -10
                        for (card in target.cards.sortCards(r.identity, handCardScore < 0)) {
                            for (moveTarget in players) {
                                val v = r.calculateMessageCardValue(fsm.whoseTurn, moveTarget, card) + handCardScore
                                if (v > value) {
                                    value = v
                                    cardId = card.id
                                    messageCardId = 0
                                    targetPlayerId = r.getAlternativeLocation(moveTarget.location)
                                }
                            }
                        }
                        for (card in target.messageCards.toList()) {
                            val removeValue = r.calculateRemoveCardValue(fsm.whoseTurn, target, card)
                            for (moveTarget in players) {
                                val v = r.calculateMessageCardValue(fsm.whoseTurn, moveTarget, card) + removeValue
                                if (v > value) {
                                    value = v
                                    cardId = 0
                                    messageCardId = card.id
                                    targetPlayerId = r.getAlternativeLocation(moveTarget.location)
                                }
                            }
                        }
                    })
                }, 3, TimeUnit.SECONDS)
            }
            return null
        }

        override fun resolveProtocol(player: Player, message: GeneratedMessage): ResolveResult? {
            if (player !== r) {
                logger.error("不是你发技能的时机")
                player.sendErrorMessage("不是你发技能的时机")
                return null
            }
            if (message !is skill_miao_shou_b_tos) {
                logger.error("错误的协议")
                player.sendErrorMessage("错误的协议")
                return null
            }
            val g = r.game!!
            if (r is HumanPlayer && !r.checkSeq(message.seq)) {
                logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${message.seq}")
                r.sendErrorMessage("操作太晚了")
                return null
            }
            if (message.cardId != 0 && message.messageCardId != 0) {
                logger.error("只能选择手牌或情报其中之一")
                player.sendErrorMessage("只能选择手牌或情报其中之一")
                return null
            }
            if (message.targetPlayerId < 0 || message.targetPlayerId >= g.players.size) {
                logger.error("目标错误")
                player.sendErrorMessage("目标错误")
                return null
            }
            val target2 = g.players[r.getAbstractLocation(message.targetPlayerId)]!!
            if (!target2.alive) {
                logger.error("目标已死亡")
                player.sendErrorMessage("目标已死亡")
                return null
            }
            val card: Card?
            if (message.cardId == 0 && message.messageCardId == 0) {
                logger.error("必须选择一张手牌或情报")
                player.sendErrorMessage("必须选择一张手牌或情报")
                return null
            } else if (message.messageCardId == 0) {
                card = target.deleteCard(message.cardId)
                if (card == null) {
                    logger.error("没有这张牌")
                    player.sendErrorMessage("没有这张牌")
                    return null
                }
            } else {
                card = target.deleteMessageCard(message.messageCardId)
                if (card == null) {
                    logger.error("没有这张牌")
                    player.sendErrorMessage("没有这张牌")
                    return null
                }
            }
            r.incrSeq()
            logger.info("${r}将${card}作为情报，面朝上移至${target2}的面前")
            g.players.send { p ->
                skillMiaoShouBToc {
                    playerId = p.getAlternativeLocation(r.location)
                    fromPlayerId = p.getAlternativeLocation(target.location)
                    targetPlayerId = p.getAlternativeLocation(target2.location)
                    if (message.cardId != 0) this.card = card.toPbCard()
                    else messageCardId = card.id
                }
            }
            return ResolveResult(
                fsm.copy(
                    messageCard = card,
                    inFrontOfWhom = target2,
                    whoseFightTurn = target2,
                    isMessageCardFaceUp = true
                ), true
            )
        }
    }

    companion object {
        fun ai(e: FightPhaseIdle, skill: ActiveSkill): Boolean {
            val player = e.whoseFightTurn
            val g = player.game!!
            !player.roleFaceUp || return false
            if (g.players.any {
                    it!!.isPartnerOrSelf(player) && it.willWin(e.whoseTurn, e.inFrontOfWhom, e.messageCard)
                }) return false
            g.players.any {
                it!!.isEnemy(player) && it.willWin(e.whoseTurn, e.inFrontOfWhom, e.messageCard)
            } || e.inFrontOfWhom.isPartnerOrSelf(player) && e.inFrontOfWhom.willDie(e.messageCard) || return false
            var target: Player? = null
            var count = 0
            fun updateMax(p: Player) {
                val c = p.cards.size + p.messageCards.size
                if (c > count) {
                    target = p
                    count = c
                }
            }
            val players = g.players.filterNotNull().filter { it.alive }.shuffled()
            if (player.identity != Black) {
                for (p in players) {
                    p.identity != player.identity && p.identity != Black || continue
                    updateMax(p)
                }
            }
            if (target == null) {
                for (p in players) {
                    p.isEnemy(player) || continue
                    updateMax(p)
                }
            }
            if (target == null) {
                for (p in players) {
                    p.isPartnerOrSelf(player) || continue
                    updateMax(p)
                }
            }
            target ?: return false
            GameExecutor.post(player.game!!, {
                skill.executeProtocol(player.game!!, player, skillMiaoShouATos {
                    targetPlayerId = player.getAlternativeLocation(target!!.location)
                })
            }, 3, TimeUnit.SECONDS)
            return true
        }
    }
}
